examples/custom_search_method/core_search_runner/asha.py

```diff
@@ -0,0 +1,483 @@
+            )
+            progress_no_overhead = num_valid_trials / self.asha_search_state.max_trials
+            progress = max(progress_no_overhead, progress)
+
+        return progress
+
+    ############################################################################
+    # User-defined function that generates hyperparameters for each trial.
+    # The hyperparameters are passed to a trial in the Create operation.
+    # In this example, the model (defined in model_def.py) is expecting the following
+    # hyperparameters:
+    #   -> global_batch_size,
+    #   -> n_filters1
+    #   -> n_filters2,
+    #   -> learning_rate,
+    #   -> dropout,
+    #   -> dropout2.
+    def sample_params(self) -> Dict[str, object]:
+        hparams = {
+            "global_batch_size": 64,
+            "n_filters1": random.randint(8, 64),
+            "n_filters2": random.randint(8, 72),
+            "learning_rate": random.uniform(0.0001, 1.0),
+            "dropout1": random.uniform(0.2, 0.8),
+            "dropout2": random.uniform(0.2, 0.8)
+        }
+        logging.info(f"hparams={hparams}")
+        return hparams
+
+    ############################################################################
+    # User-defined method for saving information related to the current state of
+    # the search. This method is invoked by CoreSearchRunner to save SearchMethod
+    # state to support fault tolerance.
+    # In this example, all information related to ASHA are stored in ASHASearchMethodState,
+    # so we use pickle.dump to save the asha_search_state object.
+    #
+    # Note: if you have multiple objects/models/variables that are used by your
+    # search method, they need to be saved in this method to enable fault tolerance.
+    def save_method_state(self, path: Path) -> None:
+        checkpoint_path = path.joinpath("method_state")
+        with checkpoint_path.open("wb") as f:
+            pickle.dump(self.asha_search_state, f)
+
+    ############################################################################
+    # User-defined method for loading SearchMethod related information.
+    # This method is invoked by CoreSearchRunner when resuming the search after
+    # the search process was terminated or interrupted.
+    # In this example, we simply load ASHASearchMethodState object that was
+    # previously saved in save_method_state().
+    #
+    # Note: to ensure that your search method is initialized correctly on
+    # resumption, load all objects/models/variables that were saved in
+    # save_method_state().
+    def load_method_state(self, path: Path) -> None:
+        checkpoint_path = path.joinpath("method_state")
+        with checkpoint_path.open("rb") as f:
+            self.asha_search_state = pickle.load(f)
+
+    ############################################################################
+    # ASHA internal implementation details. You can skip this part.
```

Comment:
let's point users to this doc if they want to understand what's happening here:
https://docs.determined.ai/latest/training/hyperparameter/search-methods/hp-adaptive-asha.html